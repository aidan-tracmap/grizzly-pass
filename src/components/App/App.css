.App {
  position: absolute;
  width: 100%;
  min-height: 100%;
}

.App-header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  height: 50px;
  z-index: 1;
}

.App-sideMenu {
  position: fixed;
  top: 50px;
  left: 0;
  bottom: 0;
  width: 60px;
  z-index: 1;
}

.App-footer {
  background: #f2f3f4;
  margin-left: 80px;
}

.App-content {
  display: flex;
  flex-direction: row;
  margin-top: 50px;
  margin-left: 60px;
  padding: 30px 30px 80px 30px;
}

.App-month {
  padding-right: 30px;
}

.App-month:last-child {
}

.App-monthTitle {
  font-size: 23px;
  font-weight: 300;
  padding-bottom: 10px;
}

.App-cardWrapper {
  padding-bottom: 10px;
}
