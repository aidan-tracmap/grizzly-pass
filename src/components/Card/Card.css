.Card {
  width: 350px;
  background: #fff;
  border: 1px solid #eee;
  box-shadow: 0 2px 3px rgba(0, 0, 0, 0.05);
  padding: 20px;
}

.Card-title {
  font-size: 20px;
  font-weight: 600;
}

.Card-person {
  font-weight: 600;
  color: #666;
}

.Card-date {
  color: #666;
}

.Card-progress {
  padding-top: 5px;
}

.Card-labels {
  display: flex;
  flex-direction: row-reverse;
}

.Card-label {
  padding-left: 4px;
}

.Card input {
  width: 100%;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-weight: inherit;
  font-size: inherit;
}
