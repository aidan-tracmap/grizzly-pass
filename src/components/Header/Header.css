.Header {
  height: 100%;
  display: flex;
  flex-direction: row;
  background: #2f5d79;
  color: white;
  box-shadow: 0 0 3px rgba(0, 0, 0, 0.1);
}

.Header-left {
  flex: 60px 0;
}

.Header-right {
  flex: 1;
  padding-left: 18px;
  display: flex;
  align-items: center;
}

.Header h1 {
  font-size: 23px;
  font-weight: 400;
  display: inline-block;
}

.Header-right-publishDate {
  display: inline-block;
  margin-left: 20px;
  font-weight: 600;
}
