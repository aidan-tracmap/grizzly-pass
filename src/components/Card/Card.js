import React, { PropTypes } from "react";
import moment from "moment";
import Selectable from "../Selectable";
import ProgressBar from "../ProgressBar";
import Label from "../Label";
import { statusIds } from "../../types";
import "./Card.css";

export default function Card(props) {
  const { title, person, time, progress, status, labels } = props;
  const date = moment(time).format("D MMMM");

  const labelsDiv = !labels
    ? null
    : <div className="Card-labels">
        {labels.map(({ id, initial, colour }) => (
          <div key={id} className="Card-label">
            <Label initial={initial} colour={colour} />
          </div>
        ))}
      </div>;

  return (
    <div className="Card">
      <div className="Card-title"><Selectable>{title}</Selectable></div>
      <div className="Card-person"><Selectable>{person}</Selectable></div>
      <div className="Card-date"><Selectable>{date}</Selectable></div>
      {labelsDiv}
      <div className="Card-progress">
        <ProgressBar progress={progress} status={status} />
      </div>
    </div>
  );
}

Card.propTypes = {
  title: PropTypes.string.isRequired,
  person: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  progress: PropTypes.number.isRequired,
  status: PropTypes.oneOf(statusIds).isRequired,
  labels: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      initial: PropTypes.string.isRequired,
      colour: PropTypes.string.isRequired
    })
  )
};
