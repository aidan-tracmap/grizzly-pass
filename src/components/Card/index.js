// @flow

import React from "react";
import moment from "moment";
import DatePicker from "../DatePicker";
import Selectable from "../Selectable";
import ProgressBar from "../ProgressBar";
import Label from "../Label";
import type { ProjectWithLabelInfo } from "../../types";
import "./Card.css";

type CardProps = {
  readonly?: boolean,
  project: ProjectWithLabelInfo,
  onProjectChange: (ProjectWithLabelInfo) => void
};

export default function Card(
  {
    readonly = true,
    project,
    onProjectChange
  }: CardProps
) {
  const { progress, status, labels } = project;

  const labelsDiv = !labels.length
    ? null
    : <div className="Card-labels">
        {labels.map(labelInfo => (
          <div key={labelInfo.id} className="Card-label">
            <Label labelInfo={labelInfo} />
          </div>
        ))}
      </div>;

  const className = ["Card", readonly && "isReadonly"].filter(f => f).join(" ");

  return (
    <div className={className}>
      <div className="Card-title">
        {renderTextElement(project, "title", readonly, onProjectChange)}
      </div>
      <div className="Card-person">
        {renderTextElement(project, "person", readonly, onProjectChange)}
      </div>
      <div className="Card-date">
        {renderDate(project, "time", readonly, onProjectChange)}
      </div>
      {labelsDiv}
      <div className="Card-progress">
        <ProgressBar progress={progress} status={status} />
      </div>
    </div>
  );
}

function renderTextElement(
  wholeProject: ProjectWithLabelInfo,
  attr: "title" | "person",
  readonly: boolean,
  onChange: (ProjectWithLabelInfo) => void
) {
  const text = wholeProject[attr];

  if (readonly) return <Selectable>{text}</Selectable>;
  else return (
      <input
        value={text}
        placeholder={attr === "title" ? "Title" : "Person"}
        onChange={e => onChange({ ...wholeProject, [attr]: e.target.value })}
      />
    );
}

function renderDate(
  wholeProject: ProjectWithLabelInfo,
  attr: "time",
  readonly: boolean,
  onChange: (ProjectWithLabelInfo) => void
) {
  const time = wholeProject[attr];
  const date = moment(time).format("D MMMM");

  if (readonly) return <Selectable>{date}</Selectable>;
  else return (
      <DatePicker
        readonly={readonly}
        time={time}
        onTimeChange={time => onChange({ ...wholeProject, time })}
      />
    );
}
