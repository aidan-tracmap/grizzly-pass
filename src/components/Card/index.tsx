import * as React from "react";
import * as moment from "moment";
import Selectable from "../Selectable";
import ProgressBar from "../ProgressBar";
import Label from "../Label";
import { Status, statusIds } from "../../types";
import "./Card.css";

interface CardProps {
  title: string;
  person: string;
  time: string;
  progress: number;
  status: Status;
  labels?: {
    id: string;
    initial: string;
    colour: string;
  }[];
}

export default function Card(props: CardProps) {
  const { title, person, time, progress, status, labels } = props;
  const date = moment(time).format("D MMMM");

  const labelsDiv = !labels
    ? null
    : <div className="Card-labels">
      {labels.map(({ id, initial, colour }) => (
        <div key={id} className="Card-label">
          <Label initial={initial} colour={colour} />
        </div>
      ))}
    </div>;

  return (
    <div className="Card">
      <div className="Card-title"><Selectable>{title}</Selectable></div>
      <div className="Card-person"><Selectable>{person}</Selectable></div>
      <div className="Card-date"><Selectable>{date}</Selectable></div>
      {labelsDiv}
      <div className="Card-progress">
        <ProgressBar progress={progress} status={status} />
      </div>
    </div>
  );
}
