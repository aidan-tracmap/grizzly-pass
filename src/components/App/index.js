// @flow

import React from "react";
import { flow } from "lodash";
import type { Dispatch } from "redux";
import { connect } from "react-redux";
import Header from "../Header";
import SideMenu from "../SideMenu";
import Card from "../Card";
import Legend from "../Legend";
import type { State } from "../../reducer";
import { getProjectIdsByMonth } from "../../reducer";
import { loadProjectJsonSuccess, loadProjectJsonFailure } from "../../actions";
import type { TabId } from "../../types";
import { importFile } from "../../import";
import { map, mapErr } from "../../result";
import "./App.css";

const importProject = (readFileAsText: (File) => Promise<string>, file: File) =>
  async (dispatch: Dispatch<*>) => {
    const asText = await readFileAsText(file);
    flow(
      importFile,
      map(fileData => dispatch(loadProjectJsonSuccess(fileData))),
      mapErr(errorMessage => dispatch(loadProjectJsonFailure(errorMessage)))
    )(asText);
  };

type AppProps = {
  projectsByMonth: { month: string, projectIds: string[] }[],
  errorMessage: ?(string | string[]),
  selectedTab: ?TabId,
  importFile: (File) => void
};

export function App(
  {
    projectsByMonth,
    errorMessage,
    selectedTab,
    importFile
  }: AppProps
) {
  const months = projectsByMonth.map(({ month, projectIds }) => {
    const cards = projectIds.map(projectId => (
      <div className="App-cardWrapper" key={projectId}>
        <Card projectId={projectId} />
      </div>
    ));
    return (
      <div key={month} className="App-month">
        <h2 className="App-monthTitle">{month}</h2>
        {cards}
      </div>
    );
  });

  const hasProjects = !!projectsByMonth.length;

  return (
    <div
      className="App"
      onDrop={e => handleDrop(importFile, e)}
      onDragOver={handleDragOver}
    >
      <div className="App-sideMenu">
        <SideMenu />
      </div>
      <div className="App-header">
        <Header />
      </div>
      <div className="App-content">
        {errorMessage != null ? formatErrors(errorMessage) : months}
      </div>
      {hasProjects &&
        errorMessage === undefined &&
        <div className="App-footer">
          <Legend />
        </div>}
    </div>
  );
}

export function mapStateToProps(state: State) {
  return {
    projectsByMonth: getProjectIdsByMonth(state)
  };
}

function mapDispatchToProps(
  dispatch: Dispatch<*>,
  { readFileAsText }: { readFileAsText: (File) => Promise<string> }
) {
  return {
    importProject: (file: File) => dispatch(importProject(readFileAsText, file))
  };
}

const AppState = connect(mapStateToProps, mapDispatchToProps)(App);
export default AppState;

function handleDragOver(e: Event) {
  e.preventDefault();
}

function handleDrop(importFile: (File) => void, e: DragEvent) {
  e.preventDefault();
  const dt = e.dataTransfer;
  if (dt && dt.items) {
    const item = Array.prototype.find.call(
      dt.items,
      ({ kind }) => kind === "file"
    );
    if (item) importFile(item.getAsFile());
  } else if (dt && dt.files.length) {
    importFile(dt.files[0]);
  }
}

function formatErrors(errorMessage: string | string[]) {
  if (typeof errorMessage === "string") return errorMessage;

  return (
    <ul>
      {errorMessage.map((message, i) => <li key={i}>{message}</li>)}
    </ul>
  );
}
