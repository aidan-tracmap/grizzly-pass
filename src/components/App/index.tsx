import * as React from "react";
import * as moment from "moment";
import Header from "../Header";
import SideMenu from "../SideMenu";
import Card from "../Card";
import Legend from "../Legend";
import { TabId, Status, statusIds, tabIds } from "../../types";
import "./App.css";

interface Project {
  id: string;
  title: string;
  person: string;
  time: string;
  progress: number;
  status: Status;
  labels?: {
    id: string;
    initial: string;
    colour: string;
    title: string;
  }[];
}

export interface AppProps {
  onFileDrop: (file: File) => void;
  title?: string;
  selectedTab?: TabId;
  onTabChange: (tabId: TabId) => void;
  errorMessage?: string | string[];
  projects?: Project[];
}

export default function App(
  {
    onFileDrop,
    title = "",
    projects = [],
    errorMessage,
    selectedTab,
    onTabChange
  }: AppProps
) {
  const months = splitIntoMonths(projects).map(({ month, projects }) => {
    const cards = projects.map(project => (
      <div className="App-cardWrapper" key={project.id}>
        <Card
          title={project.title}
          person={project.person}
          time={project.time}
          progress={project.progress}
          status={project.status}
          labels={project.labels}
        />
      </div>
    ));
    return (
      <div key={month} className="App-month">
        <h2 className="App-monthTitle">{month}</h2>
        {cards}
      </div>
    );
  });

  const hasProjects = projects && projects.length;

  return (
    <div
      className="App"
      onDrop={e => handleDrop(onFileDrop, e)}
      onDragOver={handleDragOver}
    >
      <div className="App-sideMenu">
        <SideMenu onTabChange={onTabChange} selectedTab={selectedTab} />
      </div>
      <div className="App-header">
        <Header title={title} />
      </div>
      <div className="App-content">
        {errorMessage === undefined ? months : formatErrors(errorMessage)}
      </div>
      {hasProjects &&
        errorMessage === undefined &&
        <div className="App-footer">
          <Legend projects={projects} />
        </div>}
    </div>
  );
}

function splitIntoMonths(projects: Project[]) {
  const monthIndexes = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
  return monthIndexes
    .map(monthIndex =>
      projects.filter(project => moment(project.time).month() === monthIndex))
    .filter(month => month.length > 0)
    .map(month => ({
      month: moment(month[0].time).format("MMMM"),
      projects: month
    }));
}

function handleDragOver(e: React.DragEvent<HTMLDivElement>) {
  e.preventDefault();
}

function handleDrop(onFileDrop: (file: File) => void, e: React.DragEvent<HTMLDivElement>) {
  e.preventDefault();
  const dt = e.dataTransfer;
  if (dt.items) {
    const item = Array.prototype.find.call(
      dt.items,
      ({ kind }: DataTransferItem) => kind === "file"
    );
    if (item) onFileDrop(item.getAsFile());
  } else if (dt.files.length) {
    onFileDrop(dt.files[0]);
  }
}

function formatErrors(errorMessage: string | string[]) {
  if (typeof errorMessage === "string") return errorMessage;

  return (
    <ul>
      {errorMessage.map((message, i) => <li key={i}>{message}</li>)}
    </ul>
  );
}
