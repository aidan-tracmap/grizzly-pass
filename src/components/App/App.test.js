// @flow

import type { ProjectWithLabelInfo, TabId } from "../../types";
import App from "../App";
import Card from "../Card";
import Header from "../Header";
import Legend from "../Legend";
import React from "react";
import SideMenu from "../SideMenu";
import { shallow } from "enzyme";

const projects = [
  {
    id: "1",
    title: "Coffee Swirl",
    person: "Joe Lemon",
    time: "2017-03-15T10:54:04.445Z",
    progress: 13,
    status: "ontrack",
    labels: []
  },
  {
    id: "2",
    title: "Rake Twister",
    person: "Alex Apple",
    time: "2017-04-12T10:54:04.445Z",
    progress: 50,
    status: "onhold",
    labels: [{ id: "13", initial: "A", colour: "#ff0", title: "Apple" }]
  }
];

function renderApp(
  {
    title,
    projects = [],
    handleFileDrop = jest.fn(),
    errorMessage,
    selectedTab,
    handleTabChange = jest.fn(),
    handleProjectsChange = jest.fn(),
    editable = false
  }: {
    title?: string,
    projects?: ProjectWithLabelInfo[],
    handleFileDrop?: (File) => void,
    errorMessage?: string | string[],
    selectedTab?: TabId,
    handleTabChange?: (?TabId) => void,
    handleProjectsChange?: (ProjectWithLabelInfo[]) => void,
    editable?: boolean
  } = {}
) {
  return shallow(
    <App
      projects={projects}
      title={title}
      onFileDrop={handleFileDrop}
      onTabChange={handleTabChange}
      errorMessage={errorMessage}
      selectedTab={selectedTab}
      onProjectsChange={handleProjectsChange}
      editable={editable}
    />
  );
}

const defaultApp = renderApp();

it("shows a header", () => {
  expect(defaultApp.find(Header).length).toBe(1);
});

it("passes the dummy document title to the header", () => {
  const app = renderApp({ title: "Dummy Projects 2020" });
  expect(app.find(Header).prop("title")).toBe("Dummy Projects 2020");
});

it("shows no legend when there are no projects", () => {
  expect(defaultApp.find(Legend).length).toBe(0);
});

it("strips non-label info from `projects` prop before passing it to <Legend />", () => {
  const app = renderApp({ projects });
  expect(app.find(Legend).props()).toEqual({
    projects: projects.map(({ labels }) => ({ labels }))
  });
});

it("contains no cards when passed no projects", () => {
  expect(defaultApp.find(Card).length).toBe(0);
});

it("shows a card for each project passed in", () => {
  const app = renderApp({ projects });
  expect(app.find(Card).length).toBe(projects.length);
});

it("passes project props to the <Card />", () => {
  const app = renderApp({ projects: projects.slice(0, 1) });
  expect(app.find(Card).props()).toMatchObject({
    project: {
      title: "Coffee Swirl",
      person: "Joe Lemon",
      time: "2017-03-15T10:54:04.445Z",
      progress: 13,
      status: "ontrack"
    }
  });
});

it("makes cards readonly if app isn't editable", () => {
  const app = renderApp({ projects: projects.slice(0, 1), editable: false });
  expect(app.find(Card).props()).toMatchObject({ readonly: true });
});

it("makes cards not readonly if app is editable", () => {
  const app = renderApp({ projects: projects.slice(0, 1), editable: true });
  expect(app.find(Card).props()).toMatchObject({ readonly: false });
});

it("passes labels prop to the <Card /> if it exists", () => {
  const app = renderApp({ projects: projects.slice(1, 2) });
  expect(app.find(Card).props()).toMatchObject({
    project: {
      title: "Rake Twister",
      person: "Alex Apple",
      time: "2017-04-12T10:54:04.445Z",
      progress: 50,
      status: "onhold",
      labels: [{ id: "13", initial: "A", colour: "#ff0" }]
    }
  });
});

it("splits projects into months", () => {
  const app = renderApp({ projects });
  const monthColumns = app.find(".App-month");
  expect(monthColumns.length).toBe(2);
  expect(monthColumns.at(0).find(Card).length).toBe(1);
  expect(monthColumns.at(1).find(Card).length).toBe(1);
});

it("renders month names at the top of the columns", () => {
  const app = renderApp({ projects });
  const monthColumns = app.find(".App-month .App-monthTitle");
  expect(monthColumns.at(0).text()).toBe("March");
  expect(monthColumns.at(1).text()).toBe("April");
});

it("prevents default event handling on dragover", () => {
  const preventDefault = jest.fn();
  defaultApp.simulate("dragover", { preventDefault });
  expect(preventDefault).toHaveBeenCalled();
});

it("prevents default event handling on drop", () => {
  const preventDefault = jest.fn();
  defaultApp.simulate("drop", { preventDefault, dataTransfer: { files: [] } });
  expect(preventDefault).toHaveBeenCalled();
});

it("calls drop handler when browser support the DataTransferItemList interface", () => {
  const mockFile = { MOCK: "FILE" };
  const handleFileDrop = jest.fn();
  const app = renderApp({ handleFileDrop });
  app.simulate("drop", {
    preventDefault: () => {},
    dataTransfer: {
      items: [{ kind: "file", getAsFile: () => mockFile }]
    }
  });
  expect(handleFileDrop).toHaveBeenCalledWith(mockFile);
});

it("calls drop handler when browser support the DataTransfer interface", () => {
  const mockFile = { MOCK: "FILE" };
  const handleFileDrop = jest.fn();
  const app = renderApp({ handleFileDrop });
  app.simulate("drop", {
    preventDefault: () => {},
    dataTransfer: {
      files: [mockFile]
    }
  });
  expect(handleFileDrop).toHaveBeenCalledWith(mockFile);
});

it("doesn't call drop handler it wasn't a file that was dropped", () => {
  const handleFileDrop = jest.fn();
  const app = renderApp({ handleFileDrop });
  app.simulate("drop", {
    preventDefault: () => {},
    dataTransfer: {
      items: [{ kind: "string", getAsFile: () => {} }]
    }
  });
  expect(handleFileDrop).not.toHaveBeenCalled();
});

it("displays errorMessage if it's passed as a prop", () => {
  const app = renderApp({ errorMessage: "Error Name" });
  expect(app.find(".App-content").text()).toEqual("Error Name");
});

it("displays errors as a list if errorMessage prop is an array", () => {
  const errorMessage = ["Error 1", "Error 2"];
  const app = renderApp({ errorMessage });
  expect(app.find(".App-content li").length).toBe(2);
  expect(app.find(".App-content li").at(0).text()).toBe("Error 1");
  expect(app.find(".App-content li").at(1).text()).toBe("Error 2");
});

it("passes the selectedTab to the <SideMenu />", () => {
  const app = renderApp({ selectedTab: "edit" });
  expect(app.find(SideMenu).prop("selectedTab")).toBe("edit");
});

it("calls projects change handler when only card is edited", () => {
  const handleProjectsChange = jest.fn();
  const app = renderApp({
    projects: projects.slice(0, 1),
    handleProjectsChange
  });
  const projectChanger = app.find(Card).prop("onProjectChange");
  projectChanger({
    ...projects[0],
    title: "New Title"
  });
  expect(handleProjectsChange).toHaveBeenCalledWith([
    {
      ...projects[0],
      title: "New Title"
    }
  ]);
});

it("calls projects change handler when one of many cards is edited", () => {
  const handleProjectsChange = jest.fn();
  const app = renderApp({ projects, handleProjectsChange });
  const projectChanger = app.find(Card).first().prop("onProjectChange");
  projectChanger({
    ...projects[1],
    title: "New Title"
  });
  expect(handleProjectsChange).toHaveBeenCalledWith([
    projects[0],
    {
      ...projects[1],
      title: "New Title"
    }
  ]);
});
