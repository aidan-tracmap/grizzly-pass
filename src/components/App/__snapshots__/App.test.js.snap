exports[`test can map default state to props 1`] = `
Object {
  "projectsByMonth": Array [],
}
`;
